### PR TITLE
Khalil/1899 async rpc sent tracker

### PR DESCRIPTION
**Pull request by kc1116**
This PR adds a worker pool to the gossipsub mesh tracer rpc sent tracker worker pool. RPC control messages are processed synchronously, so it's imperative that any sub-process involved in the processing of RPC's must be non-blocking. This PR also updates the tracker to keep track of the `last highest iHaves size`, this will be used during iWant flooding detection to dynamically update the sample size of iWants expected based on the most recent largest iHaves sent.

Further reading: https://github.com/dapperlabs/flow-go/issues/6472

### CODE REVIEW

File: cmd/access/node_builder/access_node_builder.go [hunk 1]
```diff
@@ -1192,12 +1192,15 @@ func (builder *FlowAccessNodeBuilder) initPublicLibp2pNode(networkKey crypto.Pri
...
-		RpcSentTrackerCacheSize:      builder.FlowConfig.NetworkConfig.GossipSubConfig.RPCSentTrackerCacheSize,
+		Logger:                                  builder.Logger,
+		Metrics:                                 networkMetrics,
+		IDProvider:                              builder.IdentityProvider,
+		LoggerInterval:                          builder.FlowConfig.NetworkConfig.GossipSubConfig.LocalMeshLogInterval,
+		RpcSentTrackerCacheCollector:            metrics.GossipSubRPCSentTrackerMetricFactory(builder.HeroCacheMetricsFactory(), network.PublicNetwork),
+		RpcSentTrackerCacheSize:                 builder.FlowConfig.NetworkConfig.GossipSubConfig.RPCSentTrackerCacheSize,
+		RpcSentTrackerWorkerQueueCacheSize:      builder.FlowConfig.NetworkConfig.GossipSubConfig.RPCSentTrackerQueueCacheSize,
+		RpcSentTrackerNumOfWorkers:              builder.FlowConfig.NetworkConfig.GossipSubConfig.RpcSentTrackerNumOfWorkers,
+		RpcSentTrackerWorkerQueueCacheCollector: metrics.GossipSubRPCSentTrackerQueueMetricFactory(builder.HeroCacheMetricsFactory(), network.PublicNetwork),
```
[the remark this answers is not in the document]
Comment (kc1116) [reply cmd/access/node_builder/access_node_builder.go]: https://github.com/onflow/flow-go/pull/4553/commits/7b52ab18decbfceb530800a79ae21b0255731e46

File: cmd/observer/node_builder/observer_builder.go [hunk 2]
```diff
@@ -703,12 +703,15 @@ func (builder *ObserverServiceBuilder) initPublicLibp2pNode(networkKey crypto.Pr
...
-		Metrics:                      builder.Metrics.Network,
-		IDProvider:                   builder.IdentityProvider,
-		LoggerInterval:               builder.FlowConfig.NetworkConfig.GossipSubConfig.LocalMeshLogInterval,
-		RpcSentTrackerCacheCollector: metrics.GossipSubRPCSentTrackerMetricFactory(builder.HeroCacheMetricsFactory(), network.PublicNetwork),
-		RpcSentTrackerCacheSize:      builder.FlowConfig.NetworkConfig.GossipSubConfig.RPCSentTrackerCacheSize,
+		Logger:                                  builder.Logger,
+		Metrics:                                 builder.Metrics.Network,
+		IDProvider:                              builder.IdentityProvider,
+		LoggerInterval:                          builder.FlowConfig.NetworkConfig.GossipSubConfig.LocalMeshLogInterval,
+		RpcSentTrackerCacheCollector:            metrics.GossipSubRPCSentTrackerMetricFactory(builder.HeroCacheMetricsFactory(), network.PublicNetwork),
```
[the remark this answers is not in the document]
Comment (kc1116) [reply cmd/observer/node_builder/observer_builder.go]: https://github.com/onflow/flow-go/pull/4553/commits/4bacc83da04645ac3950596b8c126831e9227124

File: module/component/component.go [hunk 3]
```diff
@@ -143,7 +143,8 @@ func NoopWorker(ctx irrecoverable.SignalerContext, ready ReadyFunc) {
 type ComponentManagerBuilder interface {
 	// AddWorker adds a worker routine for the ComponentManager
 	AddWorker(ComponentWorker) ComponentManagerBuilder
-
+	// AddWorkers adds n number of worker routines for the ComponentManager.
+	AddWorkers(int, ComponentWorker) ComponentManagerBuilder
```
[the remark this answers is not in the document]
Comment (kc1116) [reply module/component/component.go]: https://github.com/onflow/flow-go/pull/4553/commits/87c07e07e6d7f8804faa0a91cf1366107fa41fbc

File: network/p2p/tracer/gossipSubMeshTracer.go [hunk 4]
```diff
@@ -80,6 +95,15 @@ func NewGossipSubMeshTracer(config *GossipSubMeshTracerConfig) *GossipSubMeshTra
 			ready()
 			g.logLoop(ctx)
 		}).
+		AddWorker(func(ctx irrecoverable.SignalerContext, ready component.ReadyFunc) {
+			ready()
+			lg.Debug().Msg("starting rpc sent tracker")
```
[the remark this answers is not in the document]
Comment (kc1116) [reply network/p2p/tracer/gossipSubMeshTracer.go]: https://github.com/onflow/flow-go/pull/4553/commits/13382794a6e431fe4f6e8c4121dde4f811e39d72

File: network/p2p/tracer/internal/rpc_sent_tracker.go [hunk 5]
```diff
@@ -1,32 +1,125 @@
...
 	"github.com/onflow/flow-go/module"
+	"github.com/onflow/flow-go/module/component"
+	"github.com/onflow/flow-go/module/irrecoverable"
+	"github.com/onflow/flow-go/module/mempool/queue"
 	p2pmsg "github.com/onflow/flow-go/network/p2p/message"
 )
 
+// trackRPC is an internal data structure for "temporarily" storing *pubsub.RPC sent in the queue before they are processed
+// by the *RPCSentTracker.
+type trackRPC struct {
```
[the remark this answers is not in the document]
Comment (kc1116) [reply network/p2p/tracer/internal/rpc_sent_tracker.go]: https://github.com/onflow/flow-go/pull/4553/commits/bec095c25ce8c5715cd4e1d52932daf4bd6dfc27

File: network/p2p/tracer/internal/rpc_sent_tracker.go [hunk 6]
```diff
@@ -1,32 +1,125 @@
...
+
+	return tracker
 }
 
-// OnIHaveRPCSent caches a unique entity message ID for each message ID included in each rpc iHave control message.
+// RPCSent submits the control message to the worker queue for async tracking.
 // Args:
 // - *pubsub.RPC: the rpc sent.
-func (t *RPCSentTracker) OnIHaveRPCSent(iHaves []*pb.ControlIHave) {
+func (t *RPCSentTracker) RPCSent(rpc *pubsub.RPC) error {
```
[the remark this answers is not in the document]
Comment (kc1116) [reply network/p2p/tracer/internal/rpc_sent_tracker.go]: https://github.com/onflow/flow-go/pull/4553/commits/119f946a5b8f902eb2e30fd190cff68ba9a94752

[the remark this answers is not in the document]
Comment (kc1116) [reply network/p2p/tracer/internal/rpc_sent_tracker.go]: https://github.com/onflow/flow-go/pull/4553/commits/23c6cfe60db822d2945041faf701f1f125b4c511

File: network/p2p/tracer/internal/rpc_sent_tracker.go [hunk 7]
```diff
@@ -1,32 +1,125 @@
...
+	n, err := nonce()
+	if err != nil {
+		return fmt.Errorf("failed to get track rpc work nonce: %w", err)
+	}
+	t.workerPool.Submit(trackRPC{Nonce: n, rpc: rpc})
+	return nil
```
[the remark this answers is not in the document]
Comment (kc1116) [reply network/p2p/tracer/internal/rpc_sent_tracker.go]: https://github.com/onflow/flow-go/pull/4553/commits/e1c57c6011024d00d7bbf6e678fe3e67ebdfdf00

File: network/p2p/tracer/internal/rpc_sent_tracker.go [hunk 8]
```diff
@@ -1,32 +1,125 @@
...
+}
+
+// rpcSent tracks control messages sent in *pubsub.RPC.
+func (t *RPCSentTracker) rpcSent(work trackRPC) error {
```
[the remark this answers is not in the document]
Comment (kc1116) [reply network/p2p/tracer/internal/rpc_sent_tracker.go]: https://github.com/onflow/flow-go/pull/4553/commits/344b30f2fdf69a2f44da8245751dfa2a4c4d0906

File: network/p2p/tracer/internal/rpc_sent_tracker.go [hunk 9]
```diff
@@ -45,3 +138,38 @@ func (t *RPCSentTracker) OnIHaveRPCSent(iHaves []*pb.ControlIHave) {
...
+		case <-ctx.Done():
+			return
+		default:
+		}
+
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			t.lastHighestIHaveRPCSize.Store(0)
```
[the remark this answers is not in the document]
Comment (kc1116) [reply network/p2p/tracer/internal/rpc_sent_tracker.go]: https://github.com/onflow/flow-go/pull/4553/commits/e6a0a33b3651649bc7806ab87f51fe373b06aa5b

File: network/p2p/tracer/gossipSubMeshTracer.go [hunk 10]
```diff
@@ -155,9 +179,9 @@ func (t *GossipSubMeshTracer) Prune(p peer.ID, topic string) {
 // SendRPC is called when a RPC is sent. Currently, the GossipSubMeshTracer tracks iHave RPC messages that have been sent.
 // This function can be updated to track other control messages in the future as required.
 func (t *GossipSubMeshTracer) SendRPC(rpc *pubsub.RPC, _ peer.ID) {
-	switch {
-	case len(rpc.GetControl().GetIhave()) > 0:
-		t.rpcSentTracker.OnIHaveRPCSent(rpc.GetControl().GetIhave())
+	err := t.rpcSentTracker.RPCSent(rpc)
+	if err != nil {
+		t.logger.Err(err).Msg("failed to track sent pubsbub rpc")
```
[the remark this answers is not in the document]
Comment (kc1116) [reply network/p2p/tracer/gossipSubMeshTracer.go]: https://github.com/onflow/flow-go/pull/4553/commits/640f9bfcd9e8e7653652611fd9cd590a55fc518f

File: network/p2p/tracer/internal/rpc_sent_tracker.go [hunk 11]
```diff
@@ -1,32 +1,130 @@
...
+	// Nonce prevents deduplication in the hero store
+	Nonce []byte
+	rpc   *pubsub.RPC
+}
+
+// lastHighestIHaveRPCSize tracks the last highest rpc control message size the time stamp it was last updated.
+type lastHighestIHaveRPCSize struct {
+	*atomic.Int64
+	lastUpdate time.Time
+}
```
Comment (yhassanzadeh13) [inline network/p2p/tracer/internal/rpc_sent_tracker.go:32]: This entire structure must be concurrency-safe, currently only the size part is atomic, and the time is non-concurrency safe.

```go
// lastHighestIHaveRPCSize tracks the last highest rpc control message size the time stamp it was last updated.
type lastHighestIHaveRPCSize struct {
	sync.Mutex
	lastSize   int
	lastUpdate time.Time
}

// update updates the last highest ihave RPC size atomically.
// If the new size is larger than the current size, or if the time elapsed since the last update surpasses the reset interval.
// The time-based reset interval is used to prevent the last highest ihave RPC size from being permanently stuck at a very large size.
func (l *lastHighestIHaveRPCSize) update(size int, ttl time.Duration) {
	l.Lock()
	defer l.Unlock()

	if l.lastSize < size || time.Since(l.lastUpdate) > ttl {
		// The last highest ihave RPC size is updated if the new size is larger than the current size, or if the time elapsed since the last update surpasses the reset interval.
		l.lastSize = size
	}
}
```

File: network/p2p/tracer/internal/rpc_sent_tracker.go [hunk 12]
```diff
@@ -1,32 +1,130 @@
...
 // Args:
 // - *pubsub.RPC: the rpc sent.
-func (t *RPCSentTracker) OnIHaveRPCSent(iHaves []*pb.ControlIHave) {
+// All errors returned from this function can be considered benign.
+func (t *RPCSentTracker) Track(rpc *pubsub.RPC) error {
+	n, err := nonce()
+	if err != nil {
+		return fmt.Errorf("failed to get track rpc work nonce: %w", err)
+	}
+	if ok := t.workerPool.Submit(trackableRPC{Nonce: n, rpc: rpc}); !ok {
```
Comment (yhassanzadeh13) [inline network/p2p/tracer/internal/rpc_sent_tracker.go:100]: Please add tests for the following scenarios: 
1. Duplicate RPCs with different nonce values; both should be processed. 
2. Distinct RPCs tracked concurrently; a worker should eventually pick each.

**Review by yhassanzadeh13 — APPROVED** [review]
Looks great, except the two comments I made, the rest is good 🚀 .

File: module/component/component.go [hunk 13]
```diff
@@ -157,15 +157,23 @@ func NewComponentManagerBuilder() ComponentManagerBuilder {
...
-// within a single goroutine.
+// within a single goroutine.// AddWorker adds a ComponentWorker closure to the ComponentManagerBuilder
+
 func (c *componentManagerBuilderImpl) AddWorker(worker ComponentWorker) ComponentManagerBuilder {
 	c.workers = append(c.workers, worker)
 	return c
 }
 
+// AddWorkers adds n number of workers for the ComponentManager.
+func (c *componentManagerBuilderImpl) AddWorkers(n int, worker ComponentWorker) ComponentManagerBuilder {
```
Comment (peterargue) [inline module/component/component.go:170]: Looks like you removed this from the interface. Did you intend to revert these changes too?

File: network/p2p/tracer/internal/rpc_sent_tracker.go (hunk 11 above)
Comment (peterargue) [reply network/p2p/tracer/internal/rpc_sent_tracker.go:32]: should this also update `lastUpdate` within the lock? doesn't look like it's set anywhere

Comment (yhassanzadeh13) [reply network/p2p/tracer/internal/rpc_sent_tracker.go:32]: You are right, updated it.

Comment (kc1116) [reply network/p2p/tracer/internal/rpc_sent_tracker.go]: https://github.com/onflow/flow-go/pull/4553/commits/999eb9bdeaf15b420e76b8d76546dfd62f445611

File: network/p2p/tracer/internal/rpc_sent_tracker.go (hunk 12 above)
Comment (kc1116) [reply network/p2p/tracer/internal/rpc_sent_tracker.go:106]: https://github.com/onflow/flow-go/pull/4553/commits/7c5daef4ab8ea91ab7cb03e51a1820198a9583c5

File: module/component/component.go (hunk 13 above)
Comment (kc1116) [reply module/component/component.go]: https://github.com/onflow/flow-go/pull/4553/commits/c1d3b2e0922e62309f67db7615671e24155de1ed

File: module/component/component.go [hunk 14]
```diff
@@ -157,10 +157,10 @@ func NewComponentManagerBuilder() ComponentManagerBuilder {
 	return &componentManagerBuilderImpl{}
 }
 
-// AddWorker adds a ComponentWorker closure to the ComponentManagerBuilder
 // All worker functions will be run in parallel when the ComponentManager is started.
 // Note: AddWorker is not concurrency-safe, and should only be called on an individual builder
-// within a single goroutine.
+// within a single goroutine.// AddWorker adds a ComponentWorker closure to the ComponentManagerBuilder
```
Comment (peterargue) [inline module/component/component.go:162]: the formatting here is messed up. Can you revert the changes to this file?

**Review by peterargue — APPROVED** [review]

**Comment by kc1116** [comment]
bors merge
